src/entity.rs

```diff
@@ -282,3 +282,70 @@ impl TryFrom<DatastoreEntity> for Entity {
         })
     }
 }
+
+//
+// Datastore entity collections are used to wrap query results that may be paged
+//
+pub struct DatastoreEntityCollection {
+    entities: Vec<DatastoreEntity>,
+    query: Option<Query>,
+    end_cursor: Option<String>,
+    has_more_results: bool,
+}
+
+
+impl Default for DatastoreEntityCollection {
+    fn default() -> Self {
+        DatastoreEntityCollection {
+            entities: vec![],
+            query: None,
+            end_cursor: None,
+            has_more_results: false,
+        }
+    }
+}
+
+
+
+impl DatastoreEntityCollection {
+    pub fn from_result(
+        entities: Vec<DatastoreEntity>,
+        query: Query,
+        end_cursor: String,
+        has_more_results: bool
+    ) -> DatastoreEntityCollection {
+        DatastoreEntityCollection {
+            entities,
+            query: Some(query),
+            end_cursor: Some(end_cursor),
+            has_more_results,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ResultCollection<T> {
+    pub result: Vec<T>,
+    pub query: Option<Query>,
+    pub end_cursor: Option<String>,
+    pub has_more_results: bool,
+}
+
+impl<T> TryFrom<DatastoreEntityCollection> for ResultCollection<T> 
+where
+    T: TryFrom<DatastoreEntity>
+{
+    type Error = DatastoreParseError;
+
+    fn try_from(collection: DatastoreEntityCollection) -> Result<Self, Self::Error> {
+        let result_items: Vec<T> = collection.entities.into_iter()
+            .map(|e| e.try_into().map_err(|e| DatastoreParseError::NoProperties)) // TODO map_err? why
```

Comment:
because your type narrowing where clause states that T should implement `TryFrom`, but it does not specify WHICH `TryFrom`. It could be any `TryFrom`. And each such `TryFrom` COULD have a different `Error` bound.

But you CAN specify the associated error.

```
impl<T> TryFrom<DatastoreEntityCollection> for ResultCollection<T> 
where
    T: TryFrom<DatastoreEntity, Error = DatastoreParseError>
{
    type Error = DatastoreParseError;

    fn try_from(collection: DatastoreEntityCollection) -> Result<Self, Self::Error> {
        let result_items: Vec<T> = collection.entities.into_iter()
            .map(T::try_from)
            .collect::<Result<Vec<T>, DatastoreParseError>>()?;
        Ok(ResultCollection {
            result: result_items,
            query: collection.query,
            end_cursor: collection.end_cursor,
            has_more_results: collection.has_more_results,
        })
    }
}
```

Note that this will lead to errors in other files, but they can be fixed too.
Also note that the map can be simplified with `.map(T::try_from)`.